File: Vol.09(1800-1899)/#1842 Prime Distance/p1842.cpp
// 2916444 2008-05-15 18:54:14 Accepted 1842 C++ 00:00.09 3416K 呆滞的慢板 

// 数论题，增强的筛法，回想素数筛法
// 只要筛到最大数的开方，剩下的就是素数
// 于是这里，开一个 sqrt(2^31) 大约 65536 的素数表，然后
// 对于每个 L~U 的区间，筛掉 sqrt(U) 之内的素数的倍数即可 

#include <cstdio>
#include <cstring>

bool    isPrime[65540];
int     Prime[6544];
bool    diff[1000001];
int     top = 0,    N,  k;

int     segPrime[500000],   segTop;

int main() {
    
    memset( isPrime, 1, sizeof( isPrime ) );
    
    for( int i = 2; i < 258; ++i ) 
        if( isPrime[i] )
            for( int j = i + i; j < 65540; j += i )
                isPrime[j] = false;
    
    for( int i = 2; i < 65540; ++i )
        if( isPrime[i] )
            Prime[top++] = i;

    int     L,  U,  D;
    while( scanf( "%d%d", &L, &U ) != EOF ) {
        
        if( L == 1 )    ++L;
        D = U - L + 1;
        memset( diff, 1, D );
        
        segTop = 0;
        for( int j = 0; j < top && ( Prime[j] * Prime[j] ) <= U; ++j ) {
            k = Prime[j] - L % Prime[j];
            if( k == Prime[j] ) k = 0;
            if( k + L == Prime[j] )
                k += Prime[j];
            for( ; k <= D; k += Prime[j] )
                diff[k] = false;
        }
            
        for( int i = 0; i < D; ++i )
            if( diff[i] )   segPrime[segTop++] = L + i;

        if( segTop < 2 ) {
            puts( "There are no adjacent primes." );
            continue;
        }

        int minPos = 1, maxPos = 1, val;

        for( int i = 2; i < segTop; ++i ) {
            val = segPrime[i] - segPrime[i - 1];
            if( val < segPrime[minPos] - segPrime[minPos - 1] )
                minPos = i;
            if( val > segPrime[maxPos] - segPrime[maxPos - 1] )
                maxPos = i;
        }
        
        printf( "%d,%d are closest, %d,%d are most distant.\n",
            segPrime[minPos - 1], segPrime[minPos],
            segPrime[maxPos - 1], segPrime[maxPos] );

    }
    
}

File: 范例/Accpt/18xx/p1842.pas
PROGRAM p1842;

CONST
        MAXN = 2147483647;

VAR
        m,n                     :Longint;
        num                     :Longint;
        prime                   :array[1..4800]of Longint;
        whether                 :array[0..1000000]of Boolean;

PROCEDURE MakePrime;
var
        i,j                     :Longint;
        t                       :Longint;
begin
t:=trunc(sqrt(maxn));
fillchar(whether,sizeof(whether),1);
num:=0;
for i:=2 to t do
        if whether[i] then
                begin
                for j:=2 to t div i do whether[i*j]:=false;
                inc(num);
                prime[num]:=i;
                end;

end;

PROCEDURE Main;
var
        half                            :Longint;
        i,j,start                       :Longint;
        s                               :Longint;
        min,max                         :Longint;
        min_p1,min_p2,max_p1,max_p2     :Longint;

begin
if m=1 then m:=2;

if (n<=m) then
        begin
        writeln('There are no adjacent primes.');
        exit;
        end;

fillchar(whether,n-m+1,1);

half:=n div 2;

for i:=1 to num do
        begin
        if prime[i] > half then break;

        {if prime[3330]=0 then
                begin
                writeln(i);
                readln;
                end;}

        if m div prime[i]>1 then
                begin
                start:=m div prime[i];
                if m mod prime[i]>0 then inc(start);
                end else start:=2;

        //if i>1001 then writeln(start,' ',prime[i],' ',n div prime[i]);

        for j:=start to n div prime[i] do
                whether[prime[i] * j - m]:=false;
        end;

min:=maxint;
max:=-1;

s:=-1;
for i:=0 to n-m do
        if whether[i] then
                begin
                if s=-1 then
                        begin
                        s:=i;
                        continue;
                        end;

                if i-s>max then
                        begin
                        max:=i-s;
                        max_p1:=s;
                        max_p2:=i;
                        end;

                if i-s<min then
                        begin
                        min:=i-s;
                        min_p1:=s;
                        min_p2:=i;
                        end;
                s:=i;
                end;

if max>0 then
        writeln(min_p1+m,',',min_p2+m,' are closest, ',max_p1+m,',',max_p2+m,' are most distant.')
                else
        writeln('There are no adjacent primes.');
end;

BEGIN
MakePrime;
while not eof do
        begin
        readln(m,n);
        Main;
        end;
END.
